src/auth/mysql.rs

```diff
@@ -0,0 +1,362 @@
+#![allow(unused_qualifications)]
+//! MySql authenticator module
+use argon2rs;
+
+use diesel::prelude::*;
+use diesel::mysql::MysqlConnection;
+use diesel::result::ConnectionError;
+
+// FIXME: Remove dependency on `ring`.
+use ring::test;
+use ring::constant_time::verify_slices_are_equal;
+
+use {Error, JsonMap, JsonValue};
+use super::{AuthenticationResult, Basic};
+
+/// Error mapping for `ConnectionError`
+impl From<ConnectionError> for Error {
+    fn from(e: ConnectionError) -> Error {
+        Error::GenericError(e.to_string())
+    }
+}
+
+// Code for conversion to hex stolen from rustc-serialize:
+// https://doc.rust-lang.org/rustc-serialize/src/rustc_serialize/hex.rs.html
+
+/// MySql user record
+#[derive(Queryable)]
+pub struct User {
+    username: String,
+    pw_hash: String,
+    salt: String,
+}
+
+/// A simple authenticator that uses a MySql backed user database.
+///
+/// Requires the `mysql_authenticator` feature, which is enabled by default.
+///
+/// The user database should be a MySql database with a table of the following format:
+/// username(VARCHAR(255)), pw_hash(VARCHAR(255)), salt(VARCHAR(255))
+///
+/// # Password Hashing
+/// See `MySqlAuthenticator::hash_password` for the implementation of password hashing.
+/// The password is hashed using the [`argon2i`](https://github.com/p-h-c/phc-winner-argon2) algorithm with
+/// a randomly generated salt.
+pub struct MySqlAuthenticator {
+    database_uri: String
```

Comment:
I did have a brief look at r2d2 when adding initial diesel support, but it seemed like it wasn't able to have a `Send + Sync` `ConnectionManager` for Mysql specifically. I will happily look into it in the future though, as I agree it would be better to hold a connection pool instead of reconnecting every time we make a query. 

I should probably make sure I'm disconnecting after the query as well, until connection pooling is added.